src/System.Management.Automation/engine/parser/DynamicKeyword.cs

```diff
@@ -0,0 +1,2759 @@
+                }
+                else
+                    return allowedKeywords;
+            }
+            return null;
+        }
+    }
+
+    /// <summary>
+    /// Metadata about a member property for a dynamic keyword.
+    /// </summary>
+    public class DynamicKeywordProperty
+    {
+        /// <summary>
+        /// The name of the property
+        /// </summary>
+        public string Name { get; set; }
+
+        /// <summary>
+        /// The required type of the property
+        /// </summary>
+        public string TypeConstraint { get; set; }
+
+        /// <summary>
+        /// Any attributes that the property has
+        /// </summary>
+        public List<string> Attributes
+        {
+            get { return _attributes ?? (_attributes = new List<string>()); }
+        }
+        private List<string> _attributes;
+
+        /// <summary>
+        /// List of strings that may be used as values for this property.
+        /// </summary>
+        public List<string> Values
+        {
+            get { return _values ?? (_values = new List<string>()); }
+        }
+        private List<string> _values;
+
+        /// <summary>
+        /// Mapping the descriptive values to the actual values
+        /// </summary>
+        public Dictionary<string, string> ValueMap
+        {
+            get { return _valueMap ?? (_valueMap = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)); }
+        }
+        private Dictionary<string, string> _valueMap;
+
+        /// <summary>
+        /// Indicates that this property is mandatory and must be present.
+        /// </summary>
+        public bool Mandatory { get; set; }
+
+        /// <summary>
+        /// Indicates that this property is a key.
+        /// </summary>
+        public bool IsKey { get; set; }
+
+        /// <summary>
+        /// Indicates a range constraint on the property value
+        /// </summary>
+        public Tuple<int, int> Range { get; set; }
+    }
+
+    /// <summary>
+    /// Metadata about a parameter for a dynamic keyword. Adds one
+    /// new property to the base classL Switch for switch parameters
+    /// (THere is no such thing as a switch property...)
+    /// </summary>
+    public class DynamicKeywordParameter : DynamicKeywordProperty
+    {
+        /// <summary>
+        /// Type if this is a switch parameter and takes no argument
+        /// </summary>
+        public bool Switch { get; set; }
+    }
+
+    #endregion /* Dynamic Keyword Parser Datastructures */
+
+    #region Dynamic Keyword Metadata Reader
+
+    /// <summary>
+    /// Class to pass metadata reading errors back to the parser to be reported
+    /// </summary>
+    internal class ParseErrorContainer
+    {
+        /// <summary>
+        /// The parser string expression to invoke
+        /// </summary>
+        public Expression<Func<string>> ErrorExpr { get; }
```

Comment:
Looking at the code again, I used these types to match the `ReportError()` internal API, but I imagine calling `ReportError` using a generated `ParseError` is better. I'm also dubious about using a new type to carry parse error information out of the reader, but it seems slightly better than getting the reader to know about extents so it can construct its own `ParseError`s.

The other question is what mechanism we have to report multiple errors at once -- I think the reader best serves its purpose when it gives the user as much information as possible to correct an invalid specification.